DOCS-4447 Adds Create an API Integration Documentation (#15997)
DataDog/documentation: content/en/developers/marketplace/api_integration.md

  tag: "Documentation"
  text: "Learn about the Datadog Marketplace"
- link: "/developers/integrations/oauth_for_integrations"
  tag: "Documentation"
  text: "Learn about using OAuth for integrations"
---

## Overview

You can use [Datadog API endpoints][1] to programmatically send data to Datadog, manage your account, and visualize data collected from an API integration. 

API integrations send the following types of data to Datadog:

- [Metrics][2]
- [Logs & Log Pipelines][3]
- [Events][4]
- [Service Checks][5]
- [Traces][6]
- [Incidents][7]
- [Security Events][8]

This page provides instructions for creating an API integration in the `marketplace` repository. For more information about why you would want to create an API-based integration, see [Creating your own solution][9].

## Setup

### Prerequisites

- You must have an [API key][10] and [application key][11].
- You know which [Datadog site][12] you want to use.

An API key is required to submit data to a Datadog API endpoint. An application key is required to query data from Datadog or to create resources within the Datadog site. For more information, see [API and Application Key][11].

> Create a connection to Datadog in your company's platform using the API key, application key, and site URL, that allows users to configure and install the integration outside of Datadog.

Reviewer: The Marketplace team would like to require all API based integrations to use OAuth moving forward so this likely doesn't apply. We can replace it with the OAuth section below.

Author: ```suggestion
```

## Set up a directory and clone the Integrations Extras repository

1. Create a `dd` directory:

   {{< code-block lang="shell" >}}mkdir $HOME/dd{{< /code-block >}}
   
   The Datadog Development Toolkit command expects you to be working in the `$HOME/dd/` directory. This is not mandatory, but working in a different directory requires additional configuration steps.

2. Clone the `integrations-extras` repository:

   {{< code-block lang="shell" >}}git clone git@github.com:DataDog/integrations-extras.git{{< /code-block >}}

## Install and configure the Datadog development toolkit

The Datadog Development Toolkit command (`ddev`) allows you to create scaffolding when you are developing an integration by generating a skeleton of your integration tile's assets and metadata.

Before you begin, make sure you meet the following prerequisites:

- [Python v3.8 or later][16]
- A Python virtual environment is recommended to avoid potential environment conflicts. The instructions below use `venv`, which comes packaged with Python v3.3 and later on most operating systems.

Install and configure the development toolkit:

1. Make sure you're inside the `integrations-extras` directory:

>    {{< code-block lang="shell" >}}cd $HOME/dd/integrations-extras{{< /code-block >}}

Reviewer: Needs to be edited for the Marketplace

Author: ```suggestion
1. Make sure you're inside the `marketplace` directory:
   {{< code-block lang="shell" >}}cd $HOME/dd/marketplace{{< /code-block >}}
```

2. Set up a Python virtual environment:
   {{< code-block lang="shell" >}}
   python3 -m venv venv
   . venv/bin/activate{{< /code-block >}}

   You can exit the virtual environment at any time by running `deactivate`.

3. Install the [Developer Toolkit][17]:
   {{< code-block lang="shell" >}}pip3 install "datadog-checks-dev[cli]"{{< /code-block >}}

   If you are using the Z Shell, you may need to use escaped characters by running `pip3 install datadog-checks-dev\[cli\]`.

> 4. Set `integrations-extras` as the default working repository:
>    {{< code-block lang="shell" >}}
>    ddev config set integrations-extras $HOME/dd/integrations-extras
>    ddev config set repo integrations-extras{{< /code-block >}}

Author: ```suggestion
4. Set `marketplace` as the default working repository:
   {{< code-block lang="shell" >}}
   ddev config set marketplace $HOME/dd/marketplace
   ddev config set repo marketplace{{< /code-block >}}
```

>    If you used a directory other than `$HOME/dd` to clone the `integrations-extras` directory, use the following command to set your working repository:

Author: ```suggestion
   If you used a directory other than `$HOME/dd` to clone the `marketplace` directory, use the following command to set your working repository:
```

   {{< code-block lang="shell" >}}
   ddev config set integrations-extras <PATH/TO/INTEGRATIONS-EXTRAS>

>    ddev config set repo integrations-extras{{< /code-block >}}

Reviewer: Needs to be edited for the Marketplace

Author: ```suggestion
4. Set `integrations-extras` as the default working repository:
   {{< code-block lang="shell" >}}
   ddev config set integrations-extras $HOME/dd/integrations-extras
   ddev config set repo integrations-extras{{< /code-block >}}

   If you used a directory other than `$HOME/dd` to clone the `integrations-extras` directory, use the following command to set your working repository:

   {{< code-block lang="shell" >}}
   ddev config set marketplace <PATH/TO/MARKETPLACE>
   ddev config set repo marketplace{{< /code-block >}}
```

## Populate the integration tile scaffolding

> Run the `ddev` command to generate a skeleton of the folders and files needed for your integration. The options you use with the command are different depending on what type of integration you are developing. For a full list of the files created by the `ddev` command, see [Integrations assets][18].
> 

Reviewer: ```suggestion
Run the `ddev` command to generate a skeleton of the folders and files needed for your integration. The options you use with the command are different depending on what type of integration you are developing. For a full list of the files created by the `ddev` command, see [Integrations assets][18]. Since all the code is hosted on the partner's end, only a tile-only listing is required.

```

### Create an informational tile only listing

To create the informational tile-only listing's scaffolding:

1. Make sure you're inside the `integrations-extras` directory:

>    {{< code-block lang="shell" >}}cd $HOME/dd/integrations-extras{{< /code-block >}}

Reviewer: Needs to be updated to the Marketplace

Author: ```suggestion
1. Make sure you're inside the `marketplace` directory:
   {{< code-block lang="shell" >}}cd $HOME/dd/marketplace{{< /code-block >}}
```

Follow this template to define the `media` object in the media carousel which includes an image, a video thumbnail, and a video:

{{< code-block lang="json" filename="manifest.json" collapsible="true" >}}
"media": [
      {
        "media_type": "image",
        "caption": "A Datadog Marketplace Integration OOTB Dashboard",
        "image_url": "images/integration_name_image_name.png"
      },
      {
        "media_type": "video",
        "caption": "A Datadog Marketplace Integration Overview Video",
        "image_url": "images/integration_name_video_thumbnail.png",
        "vimeo_id": 123456789
      },
    ],
{{< /code-block >}}

For more information, see [Integrations Assets Reference][18].

## Open a pull request

Once you have access to the [`marketplace` repository][19], open a pull request that adds images (such as logos and images) and asset files (such as `Changelog.md`, `README.md`, and `manifest.json`) to your API integration's tile-only listing in the [Marketplace page][20]. Automatic tests run checks to verify that your pull request is in good shape and contains all the required content to be updated.

## Review process

> Once your pull request passes all the checks, reviewers from the `Datadog/agent-integrations` and `Datadog/documentation` teams provide suggestions and feedback on best practices.

Reviewer: Needs to be updated for the Marketplace

Author: ```suggestion
Once your pull request passes all the checks, reviewers from the `Datadog/marketplace-product-management`, `Datadog/marketplace-review`, and `Datadog/documentation` teams provide suggestions and feedback on best practices.
```

Once you have addressed the feedback and re-requested reviews, these reviewers approve your pull request.

## Coordinate GTM opportunities

Once a Marketplace tile is live, Technology Partners can meet with Datadog's Partner Marketing team to coordinate a joint go-to-market (GTM) strategy, which includes the following:

- A Datadog quote for partner press releases
- A blog post on the [Datadog Monitor][21]
- Amplification of social media posts

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: /api/latest/using-the-api/
[2]: /api/latest/metrics/
[3]: /logs/faq/partner_log_integration/
[4]: /api/latest/events/
[5]: /api/latest/service-checks/
[6]: /tracing/guide/send_traces_to_agent_by_api/
[7]: /api/latest/incidents/
[8]: /api/latest/security-monitoring/
[9]: /developers/#creating-your-own-solution

> [10]: /account_management/api-app-keys/#api-keys
> [11]: /account_management/api-app-keys/#application-keys
> [12]: /getting_started/site

Author: ```suggestion
```
